### PR TITLE
[SPARK-44806][CONNECT] Move internal client spark-connect-common to be able to test real in-process server with a real RPC client

### PR DESCRIPTION
**Pull request by juliuszsompolski**
### What changes were proposed in this pull request?

Currently, Spark Connect has the following types of integration tests:
* Client unit tests with a mock in process server (DummySparkConnectService with GRPC InProcessServerBuilder)
* Client unit tests with a mock RPC server (DummySparkConnectService with GRPC NettyServerBuilder)
* Server unit tests with in-process server and using various mocks
* E2E tests of real client with a server started in another process (using RemoteSparkSession)

What is lacking are E2E tests with an in-process Server (so that server state can be inspected asserted), and a real RPC client. This is impossible, because classes from `spark-connect-client-jvm` module include the client API which duplicates Spark SQL APIs of SparkSession, Dataset etc. When trying to pull a real client into the server module for testing, these classes clash.

Move the `org.apache.spark.sql.connect.client` code into new `spark-connect-client-jvm-internal` module, so that the internal SparkConnectClient code is separated from the client public API, and can be pulled into testing of the server.

Tried alternative approach in https://github.com/apache/spark/pull/42465. That doesn't work, because it also reorders the maven build in a way so that client is build before server, but client actually requires server to be build first to use tests with `RemoteSparkSession`
Tried alternative approach to depend on a shaded/relocated version of client in https://github.com/apache/spark/pull/42461, but that's just not possible to do neither in maven nor sbt.
Tried alternative approach to create client-jvm-internal module in https://github.com/apache/spark/pull/42441, moving things to connect-common was preferred to introducing new module by reviewers.

### Why are the changes needed?

For being able to use the internal client for testing of in-process server with an in-process client, but communicating over real RPC.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

All modules test and build.
I looked if more tests could be moved into jvm-internal, but most tests in the org.apache.spark.sql.client packages depend on the full client APIs nevertheless.

### CODE REVIEW

File: connector/connect/client/jvm/pom.xml [hunk 1]
```diff
@@ -154,6 +154,7 @@
               <include>org.checkerframework:*</include>
               <include>org.apache.spark:spark-connect-common_${scala.binary.version}</include>
               <include>org.apache.spark:spark-common-utils_${scala.binary.version}</include>
+              <include>org.apache.spark:spark-sql-api_${scala.binary.version}</include>
```
Comment (juliuszsompolski) [inline connector/connect/client/jvm/pom.xml:157]: @hvanhovell this piggyback was suggested by @LuciferYang . Looks reasonable to me.

**Comment by juliuszsompolski** [comment]
@grundprinzip @hvanhovell @LuciferYang

File: connector/connect/common/src/test/scala/org/apache/spark/sql/connect/client/util/DummySparkConnectService.scala [hunk 2]
```diff
@@ -0,0 +1,194 @@
+/*
```
Comment (juliuszsompolski) [inline connector/connect/common/src/test/scala/org/apache/spark/sql/connect/client/util/DummySparkConnectService.scala:1]: this moved from SparkConnectClientSuite

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/SparkSessionDummyIntegrationSuite.scala [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+ */
+package org.apache.spark.sql
+
+import io.grpc.{CallOptions, Channel, ClientCall, ClientInterceptor, MethodDescriptor}
+
+import org.apache.spark.sql.connect.client.SparkConnectClient
+import org.apache.spark.sql.connect.client.util.{ConnectFunSuite, DummySparkConnectService}
+
+/**
+ * Tests SparkSession integration with an RPC DummySparkConnectService.
```
Comment (juliuszsompolski) [inline connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/SparkSessionDummyIntegrationSuite.scala:25]: These two tests moved from SparkConnectClientSuite

**Comment by LuciferYang** [comment]
I need to give this a bit more thought to fully grasp the benefits of this approach. At first blush, it seems off to me to place code that's only used in the client into the common module, even though it's also used by the server module during testing. Maybe we could consider a name change for the `common` module, to better capture what it does?

**Comment by LuciferYang** [comment]
Need add 

```
<dependency>
            <groupId>com.google.guava</groupId>
            <artifactId>guava</artifactId>
            <version>${connect.guava.version}</version>
            <scope>provided</scope>
        </dependency>
        <dependency>
            <groupId>com.google.guava</groupId>
            <artifactId>failureaccess</artifactId>
            <version>${guava.failureaccess.version}</version>
            <scope>provided</scope>
        </dependency>
```
to `common/pom.xml` to avoid maven test failed

**Comment by LuciferYang** [comment]
On the other hand, we need to remember to include a testing process for the connect-common module in Github Action. Currently, these test cases that have moved to the common module haven't been executed on Github Action yet.

**Comment by juliuszsompolski** [comment]
@LuciferYang see https://github.com/apache/spark/pull/42523

> I need to give this a bit more thought to fully grasp the benefits of this approach. At first blush, it seems off to me to place code that's only used in the client into the common module, even though it's also used by the server module during testing.
> If the goal is merely to avoid adding another module, should we consider renaming the common module to more accurately reflect its purpose?

The feedback that I got was to avoid making bigger changes to module structure at this point, and that we may revisit it for 4.0. Pulling this code into server module should not hurt, as it stays benign there. Now, not trying to pull tests together, minimizes this change.

**Pull request closed without merging**